src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,164 @@
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        if (head == null) {
+            head = new Node<>(null, value, null);
+            tail = head;
+        } else {
+            Node<T> node = new Node<>(tail, value, null);
+            tail.next = node;
+            tail = node;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Invalid index: " + index);
+        }
+
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> nodeAtIndex = findNodeByIndex(index);
+            Node<T> node = new Node<>(nodeAtIndex.prev, value, nodeAtIndex);
+
+            if (nodeAtIndex.prev != null) {
+                nodeAtIndex.prev.next = node;
+            } else {
+                head = node;
+            }
+            nodeAtIndex.prev = node;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return findNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = findNodeByIndex(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> nodeAtIndex = findNodeByIndex(index);
+        unlink(nodeAtIndex);
+        return nodeAtIndex.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> node = findNodeByValue(object);
+
+        if (node != null) {
+            unlink(node);
+            return true;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Invalid index: " + index);
+        }
+
+        if (head == null) {
+            throw new NoSuchElementException("List is empty");
+        }
+
+        Node<T> node;
+        if (index <= size / 2) {
+            node = head;
+            for (int i = 0; i < index; i++) {
+                node = node.next;
+            }
+        } else {
+            node = tail;
+            for (int i = size - 1; i > index; i--) {
+                node = node.prev;
+            }
+        }
+        return node;
+    }
+
+    private Node<T> findNodeByValue(T item) {
+        Node<T> node = head;
+        while (node != null) {
+            if ((item == null && node.value == null) || (item != null && item.equals(node.value))) {
```

Comment:
```suggestion
    private Node<T> findNodeByValue(T item) {
        Node<T> node = head;
        while (node != null) {
            if (data == object || object != null && object.equals(data)) {
```
if it has same pace at heap no need to call equals() and == also checks if boths are null